clippy_lints/src/incompatible_msrv.rs

```diff
@@ -0,0 +1,140 @@
+use clippy_config::msrvs::Msrv;
+use clippy_utils::diagnostics::span_lint;
+use rustc_ast::Attribute;
+use rustc_attr::{StabilityLevel, StableSince};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::TyCtxt;
+use rustc_semver::RustcVersion;
+use rustc_session::impl_lint_pass;
+use rustc_span::def_id::DefId;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// This lint checks that no function newer than the defined MSRV (minimum
+    /// supported rust version) is used in the crate.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It would prevent this project to be actually used with the specified MSRV.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// // MSRV of 1.3.0
+    /// use std::thread::sleep;
+    /// use std::time::Duration;
+    ///
+    /// // Sleep was defined in `1.4.0`.
+    /// sleep(Duration::new(1, 0));
+    /// ```
+    ///
+    /// To fix this problem, either increase your MSRV or use another item
+    /// available in your current MSRV.
+    #[clippy::version = "1.77.0"]
+    pub INCOMPATIBLE_MSRV,
+    suspicious,
+    "ensures that all items used in the crate are available for the current MSRV"
+}
+
+pub struct IncompatibleMsrv {
+    msrv: Msrv,
+    is_above_msrv: FxHashMap<DefId, RustcVersion>,
+}
+
+impl_lint_pass!(IncompatibleMsrv => [INCOMPATIBLE_MSRV]);
+
+impl IncompatibleMsrv {
+    pub fn new(msrv: Msrv) -> Self {
+        Self {
+            msrv,
+            is_above_msrv: FxHashMap::default(),
+        }
+    }
+
+    #[allow(clippy::cast_lossless)]
+    fn get_def_id_version(&mut self, tcx: TyCtxt<'_>, def_id: DefId) -> RustcVersion {
+        if let Some(version) = self.is_above_msrv.get(&def_id) {
+            return *version;
+        }
+        let version = if let Some(version) = tcx
+            .lookup_stability(def_id)
+            .and_then(|stability| match stability.level {
+                StabilityLevel::Stable {
+                    since: StableSince::Version(version),
+                    ..
+                } => Some(RustcVersion::new(
+                    version.major as _,
+                    version.minor as _,
+                    version.patch as _,
+                )),
```

Comment:
Meow (does this work?)

```suggestion
                } => Some(version),
```

Reply:
Doesn't work unfortunately:

```
error[E0308]: `if` and `else` have incompatible types
  --> clippy_lints/src/incompatible_msrv.rs:72:16
   |
62 |            let version = if let Some(version) = tcx
   |  ________________________-
63 | |              .lookup_stability(def_id)
64 | |              .and_then(|stability| match stability.level {
65 | |                  StabilityLevel::Stable {
...  |
71 | |              version
   | |              ------- expected because of this
72 | |          } else if let Some(parent_def_id) = tcx.opt_parent(def_id) {
   | | ________________^
73 | ||             self.get_def_id_version(tcx, parent_def_id)
74 | ||         } else {
75 | ||             RustcVersion::new(1, 0, 0)
76 | ||         };
   | ||         ^
   | ||_________|
   |  |_________`if` and `else` have incompatible types
   |            expected `RustcVersion`, found `rustc_semver::RustcVersion`
   |
   = note: `rustc_semver::RustcVersion` and `RustcVersion` have similar names, but are actually distinct types
note: `rustc_semver::RustcVersion` is defined in crate `rustc_semver`
  --> /home/imperio/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustc-semver-1.1.0/src/lib.rs:58:1
   |
58 | pub enum RustcVersion {
   | ^^^^^^^^^^^^^^^^^^^^^
note: `RustcVersion` is defined in crate `rustc_session`
```